### ISSUE TITLE
How would these attributes and attribute options be expressed in another protocol?

### ISSUE THREAD

**Issue by surfnet-niels** [voperson/voperson#9]
How would these attributes and attribute options be expressed in another protocol, e.g. SAML or OIDC claims?

**Comment by boshrin** [comment]
For the pilot phase we're going to defer this, but will revisit as soon as you need this solved :)

**Comment by boshrin** [comment]
Following conversation at TNC18, Niels has agreed to with the AARC project to draft SAML and OIDC profiles that will be incorporated as companion documents.

**Comment by boshrin** [comment]
Following conversation at TNC19, members of the voPerson subcommittee discussed various options (such as encoding the attribute option into XML or JSON), but did not yet identify a specific path forward.

**Comment by skoranda** [comment]
Here is a proposal by example for SAML2.

The LDAP attribute 

`voPersonApplicationUID;app-wiki: plee@myvo.org`

would be asserted  as

```
<ns1:Attribute 
    xmlns:ns1="urn:oasis:names:tc:SAML:2.0:assertion"
    FriendlyName="voPersonApplicationUID"
    Name="https://voperson.org/voPersonApplicationUID"
    NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:uri">
    <ns1:AttributeValue>
        <ns2:voPersonApplicationUID
            xmlns:ns2="https://voperson.org">
            <ns2:app-wiki>plee@myvo.org</ns2:app-wiki>
        </ns2:voPersonApplicationUID>
    </ns1:AttributeValue>
</ns1:Attribute>
```

The multi-value LDAP attribute

```
voPersonApplicationUID;app-wiki: plee@myvo.org
voPersonApplicationUID;app-portal: PatLee
```

would be asserted as

```
<ns1:Attribute 
    xmlns:ns1="urn:oasis:names:tc:SAML:2.0:assertion"
    FriendlyName="voPersonApplicationUID"
    Name="https://voperson.org/voPersonApplicationUID"
    NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:uri">
    <ns1:AttributeValue>
        <ns2:voPersonApplicationUID
            xmlns:ns2="https://voperson.org">
            <ns2:app-wiki>plee@myvo.org</ns2:app-wiki>
            <ns2:app-portal>PatLee</ns2:app-portal>
        </ns2:voPersonApplicationUID>
    </ns1:AttributeValue>
</ns1:Attribute>
```

The LDAP attribute

`voPersonApplicationUID;app-wiki;prior: patlee@myvo.org`

would be asserted as

```
<ns1:Attribute 
    xmlns:ns1="urn:oasis:names:tc:SAML:2.0:assertion"
    FriendlyName="voPersonApplicationUID"
    Name="https://voperson.org/voPersonApplicationUID"
    NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:uri">
    <ns1:AttributeValue>
        <ns2:voPersonApplicationUID
            xmlns:ns2="https://voperson.org">
            <ns2:app-wiki>
                <ns2:prior>patlee@myvo.org</ns2:prior>
            </ns2:app-wiki>
        </ns2:voPersonApplicationUID>
    </ns1:AttributeValue>
</ns1:Attribute>
```

**Comment by skoranda** [comment]
The Shibboleth SP [appears to have the capability](https://wiki.shibboleth.net/confluence/display/SP3/DOM+AttributeDecoder) to consume the example SAML above.

**Comment by surfnet-niels** [comment]
Does this mean we would have to a namespace definition for each <ns2:app-wiki> element (and equivalent?) That would not scale very well :(

**Comment by skoranda** [comment]
No, it is not required for interoperability with the Shibboleth SP that the elements be defined, only the namespace needs to be defined. 

So if an organization like, say LIGO, has an application like gracedb, then

The LDAP attribute

`voPersonApplicationUID;app-gracedb;prior: patlee@myvo.org`

would be asserted as

```
<ns1:Attribute 
    xmlns:ns1="urn:oasis:names:tc:SAML:2.0:assertion"
    FriendlyName="voPersonApplicationUID"
    Name="https://voperson.org/voPersonApplicationUID"
    NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:uri">
    <ns1:AttributeValue>
        <ns2:voPersonApplicationUID
            xmlns:ns2="https://voperson.org">
            <ns2:app-gracedb>
                <ns2:prior>patlee@myvo.org</ns2:prior>
            </ns2:app-gracedb>
        </ns2:voPersonApplicationUID>
    </ns1:AttributeValue>
</ns1:Attribute>
```
The Shibboleth SP will be able to consume that asserted XML without anyone having to have written some formal definition of what `<https://voperson.org:app-gracedb>` means. If the Shibboleth SP is configured with

```
<Attribute name="https://voperson.org" id="voPersonApplicationUID;app-gracedb;prior">
    <AttributeDecoder xsi:type="DOMAttributeDecoder" formatter="$voPersonApplicationUID.app-gracedb.prior"/>
</Attribute>
```
then it will put into the application CGI environment the variable 

`voPersonApplicationUID;app-gracedb;prior`

with value

`patlee@myvo.org`

The only requirement is that the entity that asserts that XML (say the LIGO SAML attribute authority) and the entity that consumes it (a LIGO SP) understands what is meant by 'app-gracedb' and 'prior'.

**Comment by boshrin** [comment]
From today's subcommittee call, the proposal is to turn Scott's draft into the core of the SAML companion document.

**Comment by skoranda** [comment]
I am going to deploy a Shibboleth SP 3.x and actually prove this out.

**Comment by boshrin** [comment]
The SURF folks and I spent the morning debating the merits of various approaches. There appears to be a tension between "well structured" XML and what is supported out of the box by the Shib SP (and other SPs, for that matter). eg, @skoranda 's example above would be better formatted as

```xml
<ns1:Attribute 
    xmlns:ns1="urn:oasis:names:tc:SAML:2.0:assertion"
    xmlns:ns2="https://voperson.org"
    FriendlyName="voPersonApplicationUID"
    Name="https://voperson.org/voPersonApplicationUID"
    NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:uri">
    <ns1:AttributeValue ns2:app="gracedb" ns2:prior="true">patlee@myvo.org</ns1:AttributeValue>
</ns1:Attribute>
```

But this is not allowed by the SAML spec (SAML core 2.7.3.1), and the DOM AttributeDecoder doesn't appear to be able to parse attributes as encoded here (though it would probably be a simple RFE).

Thinking about this more generally, the complex parsing is desired to reduce the complexity for SPs in consuming specific attributes (eg: the Application UID for`gracedb` in this example). However, many (most?) deployments of voPerson are likely to involve a Proxy architecture, and so our new proposal is to solve this by supporting two models, "simple" and "advanced".

The simple model would work pretty much as eduPerson works now, by ignoring attribute options.

The advanced model would convey a voPerson-defined XML and/or JSON document, ie: it would not be constrained to SAML or Shibboleth capabilities. The expectation is that either (1) the proxy would consume this document and process it in an SP-specific fashion or (2) the entire document would be passed to the SP, which would process it as needed.

We also discussed conveying attribute definition versions, so that if (eg) a new Attribute Option were added to an attribute at some point, it would be possible to determine whether or not the SP or IdP support that particular option.

**Comment by surfnet-niels** [comment]
With the proposal for Simple underway, I propose we adopt Scot's proposal https://github.com/voperson/voperson/issues/9#issuecomment-542168196 for the complex use cases where we assume entities the need to deal with tis more complex scenario will also already be more capable in processing complex structure. 
The only thing we should test i if this complex structure is parseble at all by existing SAML software like e.g. Shib SP or SaToSa.

**Comment by boshrin** [comment]
Based on yesterday's discussion, the proxy based architecture appears to cover most identified use cases, and so passing complex structure seems to be more for edge cases that we don't have fully documented yet. For now we're going to move the advanced scenario to a future release so we can work on it more fully, probably with assistance from @surfnet-niels.
